### PR TITLE
Enable compiling with system cub.

### PR DESCRIPTION
**Pull request by trivialfis**
~- This works only with CUDA 11.4, for CUDA 11.2 there are some obscure errors with inclusive scan and custom iterators.~

- Tested with all CUDA 11.x.
- Workaround cub scan by using discard iterator in AUC.
- Limit the size of Argsort when compiled with CUDA cub.

### CODE REVIEW

**Comment by trivialfis** [comment]
Tested with 11.2 and 11.4 with and without using system cub.

File: src/metric/auc.cu [hunk 1]
```diff
@@ -331,24 +331,25 @@ float GPUMultiClassAUCOVR(common::Span<float const> predts, MetaInfo const &info
...
       thrust::make_counting_iterator(0), [=] __device__(size_t i) {
-        uint32_t class_id = i / n_samples;
-        return thrust::make_tuple(class_id, d_fptp[i].first, d_fptp[i].second);
+        return thrust::make_tuple(i, d_fptp[i].first, d_fptp[i].second);
       });
   // shrink down to pair
   auto fptp_it_out = thrust::make_transform_output_iterator(
-      dh::tbegin(d_fptp), [=] __device__(Triple const &t) {
-        return thrust::make_pair(thrust::get<1>(t), thrust::get<2>(t));
+      dh::TypedDiscard<Triple>{}, [d_fptp] __device__(Triple const &t) {
```
Comment (trivialfis) [inline src/metric/auc.cu:338]: Originally this reduces the triple to pair, cub in CUDA 11.0~11.2 doesn't handle changed type.  So here I just write the result in lambda and use discard iterator to keep the output type unchanged.

**Review by hcho3 — APPROVED** [review]

**Comment by wbo4958** [comment]
@trivialfis Could you cherry-pick this PR to 1.4.0 branch?

**Comment by trivialfis** [comment]
No, I'm planning for a new release

**Pull request merged**